packages/core/src/model.ts

```diff
@@ -0,0 +1,40 @@
+import {
+  ActionObject,
+  AssignAction,
+  Assigner,
+  PropertyAssigner,
+  assign
+} from '.';
+import { EventObject } from './types';
+
+export interface ContextModel<TC, TE extends EventObject> {
+  context: TC;
+  actions: {
+    [key: string]: ActionObject<TC, TE>;
+  };
+  assign: <TEventType extends TE['type'] = TE['type']>(
+    assigner:
+      | Assigner<TC, TE & { type: TEventType }>
```

Comment:
using `&` for extracting a concrete event type is quite horrible for the readability of the IDE hints etc, take a look:
<img width="571" alt="Screenshot 2020-12-21 at 01 13 28" src="https://user-images.githubusercontent.com/9800850/102728051-e8008380-4329-11eb-8869-aef39efea5ec.png">

I think the better results would be achieved by using `Extract` or the introduced in this PR `ExtractEvent`

Reply:
Used `ExtractEvent` 👍